### PR TITLE
WIP: Provide the means to support 1-way synchronization of files from cont…

### PR DESCRIPTION
**Pull request by netiul**
**TL;DR:**
Provides the means to support 1-way synchronization of files from container to host
 
See https://github.com/EugenMayer/docker-sync/issues/588
 - Starts an rsync daemon on the host listening on the specified port
 - Starts a container that synchorizes the volume upon fs change events using the rsync command targeting the started rsync daemon on the host

This gem includes the current changes: [docker-sync-dev.gem.zip](https://github.com/EugenMayer/docker-sync/files/2336699/docker-sync-dev.gem.zip)

**Status**
Successfully using this strategy for a week now with 
 - a sync-point with 13560 files (96M)
 - a sync-point with 16797 files (411M)
 - a sync-point with 37465 files (319M)

Startup is not slower and files are synced within seconds.

**1-way sync**
The idea is to provide a way to synchronize certain directories from **container to host**. Exactly the opposite of what the rsync-strategy does (from host to container). This can be useful for directories that are generated/changed **only** in the container.

When the `reverse_rsync`-strategy is configured, basically this is done:

- an rsync daemon is started on the host listening on the specified port by `sync_host_port`, it's configured with the docker-sync.yml `src` value as a [module](https://download.samba.org/pub/rsync/rsyncd.conf.html).
- a [zluiten/fswatch-rsync](https://hub.docker.com/r/zluiten/fswatch-rsync/) container is started with the configured volume mounted. This container watches file changes by `fswatch` and starts an rsync command upon an fs event.

**!!WARNING!!** This **will** delete files on the host that are not inside the container as this is a 1-way sync from container to host and assumes that the container is always leading.

Please note that an rsync daemon is started for each configured `reverse_rsync` container, therefore each `sync_host_port` must be unique. (This could possibly be optimized by reusing an already started rsync daemon and just add a new module, but it has some side effects.)

**docker-sync.yml**
```yaml
version: "2"

options:
  compose-file-path: 'docker-compose.yml'
  verbose: true
  max-attempt: 500
syncs:
  #IMPORTANT: ensure this name is unique and does not match your other application container name
  project-sync: # tip: add -sync and you keep consistent names as a convention
    src: ./repositories/project
    sync_strategy: 'native_osx'
    sync_userid: '33'
    sync_excludes:
      - 'vendor'
  vendor-sync: # tip: add -sync and you keep consistent names as a convention
    src: ./repositories/project/vendor
    sync_strategy: 'reverse_rsync'
    sync_userid: '33'
    sync_host_port: 20873
    sync_excludes:
      - '/some-uninteresting-vendor'
```

**TODO**
- [ ] Add unit tests
- [ ] Validate configuration upon `sync:start` (f.e. detect in use `sync_host_port` values)
- [ ] Review rsyncd pid file location?
- [ ] Add some useful code comments
- [ ] Update documentation
- [ ] Add some performance test results

### CODE REVIEW

**Comment by netiul** [comment]
@EugenMayer What's your fist impression of this feature? I've been using the way it is for quite some time now and although there are a few issues it works quite well.

**Comment by EugenMayer** [comment]
@netiul i am honest, i was not trying it and i am yet not sure how people would need such a sync in a docker-sync development environment. 

For me, docker sync is about development / devops, not some production file transfer - but maybe i have to broaden my view.

could you draw a real-life example on how to use it for what exactly? Maybe pack this as a minimum app into the boilerplate?. Thanks!

**Comment by michalkleiner** [comment]
The first use case that comes to my head for this would node_modules installed inside a node container. Generally you don't need them outside on the host and they are excluded from the sync, but having them on the host 1-way synced from the container would help IDEs to work with them for code hinting etc.

**Comment by matthew-gill** [comment]
Biggest use case for me would be during symfony development. Syncing the cache back from container to host one way only would be really nice!

**Comment by EugenMayer** [comment]
Well thank you guys, interestingly i should have noticed this myself. I am already using exactly that, stacks where i skip docker sync since:
 - i do not sync vendor/node_modules/.m2 - i run `composer install` `npm install` `mvn install` on the container
 - i do not sync the framework files, they are part of the installed symbols in step one
 - i only mount the folder with the 1-5 contrib modules, depending on the framework, i might have to symlink the folder in the container
 - when that startup-script finished, i run something like `copy_symbols.sh` on the host, which then copies vendor/node_modules/.m2 / whatever symbol folder to my local folder so that my IDE has symbol support - which is absolutely critical for me

The latter step is done manually, once running a script. Honestly until now i consider this "ok" due to:
 - one does rarely change dependencies, so rerunning `composer/npm/mvn install` is rare, so resyncing the symbol folders is a rare task
 - so running all this like that

start.sh
```bash
docker-compose up -d
docker exec app composer install
./sync_symbolsh.sh
```

sync_symbolsh.sh
```bash
rm -fr vendor
docker cp app:/var/www/vendor ./
```

----
so that said until know i had the same setup you all mention, just using manual scripts since this is not "code sync", syncing that once a week seems to be already good enough mostly.

Still i understand, that can be automated with docker-sync .. and with little costs:

pros:
 - sync runs entirely on a linux filesystem with linux filewatchers -> no hickups, no pain
 - far less CPU usage then using the inverse setup, less hickups

cons:
 - watching a nodes folder for changes means watchen 30.000 files, that still harms the CPU (for some amount)

---

So my question is, does the pro really overweight the con for a an actual sync which happens so rarely? 

Thank you for sharing your stories!

**Comment by matthew-gill** [comment]
Your setup sounds similar to mine. I sync my src files, but exclude PHP vendor and node modules from syncing. I run a script every now and again to do a docker copy of those  third party dependencies back to host.

My only use case would be for symfony with cache writes only performed on the container. Having the cache up to date on the host is useful for debugging (also the PHPStorm module for symfony uses the cache files).

I'd say I'm still keen for this feature!

**Comment by michalkleiner** [comment]
We actually change npm dependencies quite often when developing on different branches with different dev deps and only merge them once the feature is finished, so part of our workflow, after switching branches, is to run `npm install` to get what the latest lock file prescribes. Potentially same for PHP deps installed via composer.
1-way sync from the container to the host would help us and if it was driven by FS events rather than by watching files, the CPU footprint might be quite small I'd think. Our host env is MacOS.

**Comment by netiul** [comment]
@EugenMayer 
> For me, docker sync is about development / devops, not some production file transfer - but maybe i have to broaden my view.

Don't worry, I used it for development only as well. I wouldn't consider this for production at all.

> could you draw a real-life example on how to use it for what exactly? Maybe pack this as a minimum app into the boilerplate?. Thanks!

I noticed nobody tried the code in this pull request yet, so I guess it's not very clear how it works. I created some real life example using the Symfony Demo app. I pushed it here: https://github.com/netiul/docker-sync-reverse-rsync-symfony-boilerplate

The example mounts the code using the native_osx strategy for syncing, but excludes the `var`, `node_modules` and `vendor` folders. Those folders are separately synced using this `reverse_rsync` strategy. The [README](https://github.com/netiul/docker-sync-reverse-rsync-symfony-boilerplate/blob/master/README.md) contains setup instructions to see it all in action.

A few issues I encouter:

 * One is that with much changes, like a fresh `composer install` or `npm install` the file system event queue will overflow, causing a delay in syncing. How much delay depends, but is max a minute, most of the time less. You can see it happening in this screen recording: https://youtu.be/rgsTsIPaqBM. I find the delay not much of a problem, because these changes are not required to be available with seconds.

 * Another issue is that event queue overflows happen without the user actual performing changes in the folders. I'm not sure why this is happening, but looks like is one reason the strategy is CPU intensive. This needs some more investigation.

> cons:
> 
>  *   watching a nodes folder for changes means watchen 30.000 files, that still harms the CPU (for some amount)

This is true. I've seen in happening with the changes I propose, but this might be less noticeable on smaller projects that I've been using it. That's why I'm curious about other users trying the code.
It also might be less issue if I can fix the occurrence of the sudden event queue overflows. (See previous point.)

**Comment by netiul** [comment]
I'm closing this branch because I'm not happy with the solution as it is and because there's no feedback.

**Comment by EugenMayer** [comment]
@netiul thats a pitty, because i think there has been a really fruitful discussion here about the use cases and that is def. worth it implementing it.

I understand you might be frustrated that there is not a lot of feedback, but i am just too busy with other things and rely on more people jumping it, especially the ones actually going to use it.

Maybe you reconsider your decision. Are there any technical reasons you had to ditch this plan?

**Comment by netiul** [comment]
@EugenMayer I guess I can reopen for the discussion that's already going on here. Altough the issue would be a better place for it.

The reason I'm not happy with the solution is that it's just not performing well enough. When using this `reverse-rsync` strategy in a setup as mentioned in the top post I noticed that it starts to affect the `native-osx` strategy container syncing the actual dev code.

Even when setting the soft and hard open-file limits of on the host machine to 500,000 and 1,000,000 respectively.

**Comment by EugenMayer** [comment]
@netiul thank you for the elaborated answer - i had similar issues / results and that is why i do not use anything with those "watchers".

The issue is not native_osx, but rather OSXFS and the way the FS events stack/propaget. The more events you feed in there the more you get into trouble, be it file-number related or just "more syncs".

Let me know what we should do with this branch - or anybody else. If nobody is really interested in this, i would rather not merge it, since this means there is no maintenance for this

**Comment by EugenMayer** [comment]
understood that this is actually problematic and might just not to be a good, solid way to solve this kind of issue.

Closing the PR now, even though i see so much work already done, as i see also the author went away from it
